It looks like our pipeline is failing to build `gingko`:

```
1820==> [2021-08-25-17:47:16.434486] Error: ProcessError: Command exited with status 2:
1821    'make' '-j16' 'test_install'
18224 errors found in build log:
1823     69387    make[4]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1824              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp/te
1825              st/test_install'
1826     69388    cd /tmp/root/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp3fbvop4k
1827              iriboyk2tf4xnbwlb/spack-build-3fnjzlp/test/test_install && /tmp/r
1828              oot/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp3fbvop4kiriboyk2t
1829              f4xnbwlb/spack-build-3fnjzlp/test/test_install//test_install
1830     69389    terminate called after throwing an instance of 'gko::CudaError'
1831     69390      what():  /tmp/root/spack-stage/spack-stage-ginkgo-1.4.0-3fnjzlp
1832              3fbvop4kiriboyk2tf4xnbwlb/spack-src/cuda/base/executor.cpp:223: g
1833              et_num_devices: cudaErrorInsufficientDriver: CUDA driver version 
1834              is insufficient for CUDA runtime version
1835     69391    Aborted (core dumped)
1836     69392    CMakeFiles/test_install.dir/build.make:73: recipe for target 'CMa
1837              keFiles/test_install' failed
1838  >> 69393    make[3]: *** [CMakeFiles/test_install] Error 134
1839     69394    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1840              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1841     69395    CMakeFiles/Makefile2:351: recipe for target 'CMakeFiles/test_inst
1842              all.dir/all' failed
1843  >> 69396    make[2]: *** [CMakeFiles/test_install.dir/all] Error 2
1844     69397    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1845              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1846     69398    CMakeFiles/Makefile2:358: recipe for target 'CMakeFiles/test_inst
1847              all.dir/rule' failed
1848  >> 69399    make[1]: *** [CMakeFiles/test_install.dir/rule] Error 2
1849     69400    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-gin
1850              kgo-1.4.0-3fnjzlp3fbvop4kiriboyk2tf4xnbwlb/spack-build-3fnjzlp'
1851     69401    Makefile:192: recipe for target 'test_install' failed
1852  >> 69402    make: *** [test_install] Error 2
```

Does that suggest a fix to you?

--- a/var/spack/repos/builtin/packages/ginkgo/1.4.0_skip_invalid_smoke_tests.patch
+++ b/var/spack/repos/builtin/packages/ginkgo/1.4.0_skip_invalid_smoke_tests.patch
@@ -1,0 +1,95 @@
+diff --git a/test/test_install/test_install.cpp b/test/test_install/test_install.cpp
+index fb9f479c06..cd2ac76855 100644
+--- a/test/test_install/test_install.cpp
++++ b/test/test_install/test_install.cpp
+@@ -35,6 +35,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ 
+ #include <chrono>
+ #include <cmath>
++#include <cstdlib>
+ #include <fstream>
+ #include <iostream>
+ #include <map>
+@@ -93,8 +94,7 @@ void check_solver(std::shared_ptr<gko::Executor> exec,
+                   const gko::matrix::Dense<> *b, gko::matrix::Dense<> *x)
+ {
+     using Mtx = gko::matrix::Csr<>;
+-    auto A =
+-        gko::share(Mtx::create(exec, std::make_shared<Mtx::load_balance>()));
++    auto A = gko::share(Mtx::create(exec, std::make_shared<Mtx::classical>()));
+ 
+     auto num_iters = 20u;
+     double reduction_factor = 1e-7;
+@@ -117,8 +117,8 @@ void check_solver(std::shared_ptr<gko::Executor> exec,
+ #if defined(HAS_HIP) || defined(HAS_CUDA)
+     // If we are on a device, we need to run a reference test to compare against
+     auto exec_ref = exec->get_master();
+-    auto A_ref = gko::share(
+-        Mtx::create(exec_ref, std::make_shared<Mtx::load_balance>()));
++    auto A_ref =
++        gko::share(Mtx::create(exec_ref, std::make_shared<Mtx::classical>()));
+     A_ref->read(A_raw);
+     auto solver_gen_ref =
+         Solver::build()
+@@ -147,13 +147,35 @@ class PolymorphicObjectTest : public gko::PolymorphicObject {};
+ int main()
+ {
+ #if defined(HAS_CUDA)
+-    auto exec = gko::CudaExecutor::create(0, gko::ReferenceExecutor::create());
++    auto extra_info = "(CUDA)";
++    using exec_type = gko::CudaExecutor;
+ #elif defined(HAS_HIP)
+-    auto exec = gko::HipExecutor::create(0, gko::ReferenceExecutor::create());
++    auto extra_info = "(HIP)";
++    using exec_type = gko::HipExecutor;
+ #else
+-    auto exec = gko::ReferenceExecutor::create();
++    auto extra_info = "(REFERENCE)";
++    using exec_type = gko::ReferenceExecutor;
+ #endif
+ 
++    std::shared_ptr<exec_type> exec;
++    try {
++#if defined(HAS_CUDA) || defined(HAS_HIP)
++        exec = exec_type::create(0, gko::ReferenceExecutor::create());
++#else
++        exec = exec_type::create();
++#endif
++        // We also try to to synchronize to ensure we really have an available
++        // device
++        exec->synchronize();
++    } catch (gko::Error &e) {
++        // Exit gracefully to not trigger CI errors. We only skip the tests in
++        // this setting
++        std::cerr
++            << "test_install" << extra_info
++            << ": a compatible device could not be found. Skipping test.\n";
++        std::exit(0);
++    }
++
+     using vec = gko::matrix::Dense<>;
+ #if HAS_REFERENCE
+     auto b = gko::read<vec>(std::ifstream("data/b.mtx"), exec);
+@@ -329,7 +351,7 @@ int main()
+     // core/matrix/csr.hpp
+     {
+         using Mtx = gko::matrix::Csr<>;
+-        auto test = Mtx::create(exec, std::make_shared<Mtx::load_balance>());
++        auto test = Mtx::create(exec, std::make_shared<Mtx::classical>());
+     }
+ 
+     // core/matrix/dense.hpp
+@@ -491,13 +513,6 @@ int main()
+                 .with_criteria(std::move(time), std::move(iteration))
+                 .on(exec);
+     }
+-#if defined(HAS_CUDA)
+-    auto extra_info = "(CUDA)";
+-#elif defined(HAS_HIP)
+-    auto extra_info = "(HIP)";
+-#else
+-    auto extra_info = "(REFERENCE)";
+-#endif
+     std::cout << "test_install" << extra_info
+               << ": the Ginkgo installation was correctly detected "
+                  "and is complete."

--- a/var/spack/repos/builtin/packages/ginkgo/package.py
+++ b/var/spack/repos/builtin/packages/ginkgo/package.py
@@ -124,8 +124,8 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
     @run_after('install')
     def setup_build_tests(self):
         """Build and install the smoke tests."""
-        # For now only develop and next releases support this scheme.
-        if not self.spec.satisfies('@develop') and not self.spec.satisfies('@1.4.0:'):
+        # For now only 1.4.0 and later releases support this scheme.
+        if self.spec.satisfies('@:1.3.0'):
             return
         with working_dir(self.build_directory):
             make("test_install")
@@ -135,8 +135,8 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
 
     def test(self):
         """Run the smoke tests."""
-        # For now only develop and next releases support this scheme.
-        if not self.spec.satisfies('@develop') and not self.spec.satisfies('@1.4.0:'):
+        # For now only 1.4.0 and later releases support this scheme.
+        if self.spec.satisfies('@:1.3.0'):
             print("SKIPPED: smoke tests not supported with this Ginkgo version.")
             return
         files = [('test_install', [r'REFERENCE',

--- a/var/spack/repos/builtin/packages/ginkgo/package.py
+++ b/var/spack/repos/builtin/packages/ginkgo/package.py
@@ -19,6 +19,7 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
 
     version('develop', branch='develop')
     version('master', branch='master')
+    version('1.4.0', commit='f811917c1def4d0fcd8db3fe5c948ce13409e28e')  # v1.4.0
     version('1.3.0', commit='4678668c66f634169def81620a85c9a20b7cec78')  # v1.3.0
     version('1.2.0', commit='b4be2be961fd5db45c3d02b5e004d73550722e31')  # v1.2.0
     version('1.1.1', commit='08d2c5200d3c78015ac8a4fd488bafe1e4240cf5')  # v1.1.1
@@ -52,13 +53,9 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
 
     # ROCm 4.1.0 breaks platform settings which breaks Ginkgo's HIP support.
     conflicts("^hip@4.1.0:", when="@:1.3.0")
-    conflicts("^hip@4.1.0:", when="@master")
     conflicts("^hipblas@4.1.0:", when="@:1.3.0")
-    conflicts("^hipblas@4.1.0:", when="@master")
     conflicts("^hipsparse@4.1.0:", when="@:1.3.0")
-    conflicts("^hipsparse@4.1.0:", when="@master")
     conflicts("^rocthrust@4.1.0:", when="@:1.3.0")
-    conflicts("^rocthrust@4.1.0:", when="@master")
 
     def cmake_args(self):
         # Check that the have the correct C++ standard is available

--- a/var/spack/repos/builtin/packages/ginkgo/package.py
+++ b/var/spack/repos/builtin/packages/ginkgo/package.py
@@ -57,6 +57,10 @@ class Ginkgo(CMakePackage, CudaPackage, ROCmPackage):
     conflicts("^hipsparse@4.1.0:", when="@:1.3.0")
     conflicts("^rocthrust@4.1.0:", when="@:1.3.0")
 
+    # Skip smoke tests if compatible hardware isn't found
+    patch('1.4.0_skip_invalid_smoke_tests.patch', when='@master')
+    patch('1.4.0_skip_invalid_smoke_tests.patch', when='@1.4.0')
+
     def cmake_args(self):
         # Check that the have the correct C++ standard is available
         if self.spec.satisfies('@:1.2.0'):

